protocol/encryption/persistence.go

```diff
@@ -766,63 +767,157 @@ func (s *sqlitePersistence) GetHashRatchetKeyByID(groupID []byte, keyID uint32,
 	var key, hash []byte
 	var seqNoPtr *uint32
 
-	err = stmt.QueryRow(groupID, keyID, seqNo).Scan(&key, &seqNoPtr, &hash)
+	oldFormat := ratchet.IsOldFormat()
+	if oldFormat {
+		// Query using the deprecated format
+		err = stmt.QueryRow(ratchet.GroupID, nil, seqNo, ratchet.DeprecatedKeyID()).Scan(&key, &seqNoPtr, &hash)
+
+	} else {
+		keyID, err := ratchet.GetKeyID()
+		if err != nil {
+			return nil, err
+		}
+
+		err = stmt.QueryRow(ratchet.GroupID, keyID, seqNo, ratchet.DeprecatedKeyID()).Scan(&key, &seqNoPtr, &hash)
+		if err != nil {
+			return nil, err
+		}
+	}
 	var seqNoResult uint32
 	if seqNoPtr == nil {
 		seqNoResult = 0
 	} else {
 		seqNoResult = *seqNoPtr
 	}
 
-	res := &HRCache{
-		KeyID: keyID,
-		Key:   key,
-		Hash:  hash,
-		SeqNo: seqNoResult,
-	}
 	switch err {
 	case sql.ErrNoRows:
 		return nil, nil
 	case nil:
+		ratchet.Key = key
+		keyID, err := ratchet.GetKeyID()
+
+		if err != nil {
+			return nil, err
+		}
+
+		res := &HRCache{
+			KeyID: keyID,
+			Key:   key,
+			Hash:  hash,
+			SeqNo: seqNoResult,
+		}
+
 		return res, nil
 	default:
 		return nil, err
 	}
 }
 
+type HashRatchetKeyCompatibility struct {
+	GroupID   []byte
+	keyID     []byte
+	Timestamp uint64
+	Key       []byte
+}
+
+func (h *HashRatchetKeyCompatibility) DeprecatedKeyID() uint32 {
+	return uint32(h.Timestamp)
+}
+
+func (h *HashRatchetKeyCompatibility) IsOldFormat() bool {
+	return len(h.keyID) == 0 && len(h.Key) == 0
+}
+
+func (h *HashRatchetKeyCompatibility) GetKeyID() ([]byte, error) {
+	if len(h.keyID) != 0 {
+		return h.keyID, nil
+	}
+
+	if len(h.GroupID) == 0 || h.Timestamp == 0 || len(h.Key) == 0 {
+		return nil, errors.New("could not create key")
+	}
+
+	return generateHashRatchetKeyID(h.GroupID, h.Timestamp, h.Key), nil
+}
+
+func (h *HashRatchetKeyCompatibility) GenerateNext() (*HashRatchetKeyCompatibility, error) {
+
+	ratchet := &HashRatchetKeyCompatibility{
+		GroupID: h.GroupID,
+	}
+
+	// Randomly generate a hash ratchet key
+	hrKey, err := crypto.GenerateKey()
+	if err != nil {
+		return nil, err
+	}
+	hrKeyBytes := crypto.FromECDSA(hrKey)
+
+	if err != nil {
+		return nil, err
+	}
+
+	currentTime := GetCurrentTime()
+	if h.Timestamp < currentTime {
+		ratchet.Timestamp = bumpKeyID(currentTime)
```

Comment:
I am wondering why do we bump here?

Reply:
I can't remember,  I guess we want it to bump of at least one, but not sure, it would make sense to bump of a random interval to avoid conflicts, but after this change conflicts are not possible/not damaging